src/main/scala/io/iohk/ethereum/jsonrpc/server/http/RateLimit.scala

```diff
@@ -1,25 +1,43 @@
 package io.iohk.ethereum.jsonrpc.server.http
 
-import java.time.Clock
-
-import akka.http.scaladsl.model.RemoteAddress
-import com.twitter.util.LruMap
-import io.iohk.ethereum.jsonrpc.server.http.JsonRpcHttpServer.JsonRpcHttpServerConfig
-
-trait RateLimit {
-
-  val config: JsonRpcHttpServerConfig
-
-  val latestRequestTimestamps = new LruMap[RemoteAddress, Long](config.rateLimit.latestTimestampCacheSize)
-
-  val clock: Clock = Clock.systemUTC()
+import akka.http.scaladsl.model.{RemoteAddress, StatusCodes}
+import akka.http.scaladsl.server.{Directive0, Route}
+import io.iohk.ethereum.db.cache.SimpleLRU
+import io.iohk.ethereum.jsonrpc.server.http.JsonRpcHttpServer.RateLimitConfig
+import akka.http.scaladsl.server.Directives._
+import io.iohk.ethereum.jsonrpc.JsonRpcError
+import de.heikoseeberger.akkahttpjson4s.Json4sSupport
+import io.iohk.ethereum.jsonrpc.serialization.JsonSerializers
+import org.json4s.{DefaultFormats, Formats, Serialization, native}
+
+class RateLimit(config: RateLimitConfig) extends Directive0 with Json4sSupport {
+
+  private implicit val serialization: Serialization = native.Serialization
+  private implicit val formats: Formats = DefaultFormats + JsonSerializers.RpcErrorJsonSerializer
+
+  lazy val lru = new SimpleLRU[RemoteAddress](
+    config.latestTimestampCacheSize,
+    config.minRequestInterval.toMillis
+  )
+
+  // Such algebras prevent if-elseif-else boilerplate in the JsonRPCServer code
+  // It is also guaranteed that:
+  //   1) config.enabled is checked only once - on route init
+  //   2) no IP address is extracted unless config.enabled is true
+  //   3) no LRU is created unless config.enabled is true
+  val rateLimitAlgebra: (Unit => Route) => Route = {
```

Comment:
This looks like premature optimization to me, I would leave a `def` that reads the `config.enabled` value each time. But if we're going to do this, we could have a `LimitStrategy` trait with two implementations: `RateLimit` (extract IP, have the LRU cache instantiated, limit if necessary) and `NoLimit` (just a no-op). The `JsonRpcHttpServer` would read the config once and instantiate the corresponding `LimitStrategy`. What do you think?

Reply:
Fixed that.